proposals/nnnn-nil-values-for-objective-c-primitives.md

```diff
@@ -0,0 +1,144 @@
+
+## Introduction
+
+This proposal aims to improve Swift’s compatibility with Objective-C by allowing Objective-C methods to declare nullability for primitives by using sentinel values. By using the format `nullable(value)`, the syntax will stay close to existing Objective-C nullability specifiers, as well as serving as inline documentation about the method’s behavior.
+
+
+Swift-evolution thread: [Pitch Discussion](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20170213/031928.html)
+
+## Motivation
+
+Many Objective-C methods use sentinel values to represent null values, whether it’s `NSNotFound`, an `NSRange` with `NSNotFound` as the location, or even simply returning `-1` for an `NSInteger`. When these methods are used from within Swift, the Swift developer must know the method’s behavior and account for these values in their code. Depending on how well-documented the code is, it can be very easy to miss these values and cause problems down the line.
+
+## Proposed solution
+
+A good example is `NSArray`’s `‑indexOfObject:` method, which could be annotated thusly:
+
+```Objective-C
+- (nullable(NSNotFound) NSUInteger)indexOfObject:(ObjectType)anObject;
+```
+
+This annotation does two things: first, it documents the behavior that a null value will return as `NSNotFound`, and second, it allows the Swift compiler to convert it to this Swift equivalent:
+
+```Swift
+func index(of object: Any) -> Int?
+```
+
+Another example is `NSPropertyListSerialization`’s `‑writePropertyList:toStream:format:options:error:` method. This method’s full signature is as follows:
+
+```Objective-C
++ (NSInteger)writePropertyList:(id)plist
+                      toStream:(NSOutputStream *)stream
+                        format:(NSPropertyListFormat)format
+                       options:(NSPropertyListWriteOptions)opt
+                         error:(out NSError * _Nullable *)error;
+```
+
+Notice that the method returns an `NSInteger`. The very last line of the documentation reads:
+
+> Returns the number of bytes written to the stream. If the value is 0 an error occurred.
+
+This is easy to miss. An annotated version would read as follows:
+
+```Objective-C
++ (nullable(0) NSInteger)writePropertyList:(id)plist
+                                  toStream:(NSOutputStream *)stream
+                                    format:(NSPropertyListFormat)format
+                                   options:(NSPropertyListWriteOptions)opt
+                                     error:(out NSError * _Nullable *)error;
+```
+
+Here, the first thing the developer sees is that `0` is a null value upon return, and the Swift compiler will treat it as an `Int`, forcing the developer to handle `nil` cases.
+
+Another place where primitives are commonly used is CoreGraphics. Here, we could annotate a method on `CGContext` to return the current point in a non-empty path:
+
+```Objective-C
+nullable(CGPointZero) CGPoint CGContextGetPathCurrentPoint(CGContextRef c);
```

Comment:
Not sure this is a good example, as there could be a valid current point that is (0,0) in an existing path. Then you have a valid value mistakenly converted to `nil`